### PR TITLE
Trace storage blob

### PR DESCRIPTION
**Pull request by SuyogSoti**
For reviewers, there are two things to look for.

1. I add the `@distributed_trace` or `@distributed_trace_async` for every function that will lead to a network call.
1. I have added the DistributedTracingPolicy to the correct place.
1. I propagate the context with `tracing_context.with_current_context` any time a new thread is created. This should work with both `threading` and `ThreadPoolExecutor`. An example is the following:

`future = executor.submit(tracing_context.with_current_context(uploader.process_chunk), chunk)`

### CODE REVIEW

**Comment by adxsdk6** [comment]
Can one of the admins verify this patch?

File: sdk/storage/azure-storage-blob/azure/storage/blob/_shared/utils.py [hunk 1]
```diff
@@ -0,0 +1,608 @@
+# -------------------------------------------------------------------------
```
Comment (annatisch) [inline sdk/storage/azure-storage-blob/azure/storage/blob/_shared/utils.py]: This file has been refactored away. Did you make an edits to it? Or is it just here as merge artifact?

File: sdk/storage/azure-storage-blob/azure/storage/blob/blob_client.py [hunk 2]
```diff
@@ -221,6 +222,7 @@ def from_connection_string(
         return cls(
             account_url, container=container, blob=blob, snapshot=snapshot, credential=credential, **kwargs)
 
+    @distributed_trace
```
Comment (annatisch) [inline sdk/storage/azure-storage-blob/azure/storage/blob/blob_client.py]: I don't think this function does a network call.

File: sdk/storage/azure-storage-blob/azure/storage/blob/blob_service_client.py [hunk 3]
```diff
@@ -149,6 +150,7 @@ def from_connection_string(
             kwargs['secondary_hostname'] = secondary
         return cls(account_url, credential=credential, **kwargs)
 
+    @distributed_trace
     def generate_shared_access_signature(
```
Comment (annatisch) [inline sdk/storage/azure-storage-blob/azure/storage/blob/blob_service_client.py]: Function doesn't do a network call.

File: sdk/storage/azure-storage-blob/azure/storage/blob/container_client.py [hunk 4]
```diff
@@ -180,6 +181,7 @@ def from_connection_string(
         return cls(
             account_url, container=container, credential=credential, **kwargs)
 
+    @distributed_trace
     def generate_shared_access_signature(
```
Comment (annatisch) [inline sdk/storage/azure-storage-blob/azure/storage/blob/container_client.py]: Function doesn't do a network call.

File: sdk/storage/azure-storage-blob/azure/storage/blob/container_client.py [hunk 5]
```diff
@@ -375,6 +379,7 @@ def delete_container(
         except StorageErrorException as error:
             process_storage_error(error)
 
+    @distributed_trace
     def acquire_lease(
```
Comment (annatisch) [inline sdk/storage/azure-storage-blob/azure/storage/blob/container_client.py]: This function calls into the Lease.aquire function which you have decorated separately - does that matter?

(This comment applies to a few different places. I wont comment on every one unless you let me know if it's an issue, in which case I can let you know all the places it applies)

File: sdk/storage/azure-storage-blob/azure/storage/blob/container_client.py [hunk 6]
```diff
@@ -747,6 +759,7 @@ def walk_blobs(
             marker=marker,
             delimiter=delimiter)
 
+    @distributed_trace
     def upload_blob(
```
Comment (annatisch) [inline sdk/storage/azure-storage-blob/azure/storage/blob/container_client.py]: This function calls into the BlobClient.upload function which you have decorated separately - does that matter?

File: sdk/storage/azure-storage-blob/azure/storage/blob/container_client.py [hunk 7]
```diff
@@ -655,6 +665,7 @@ def set_container_access_policy(
         except StorageErrorException as error:
             process_storage_error(error)
 
+    @distributed_trace
     def list_blobs(self, name_starts_with=None, include=None, marker=None, timeout=None, **kwargs):
```
Comment (annatisch) [inline sdk/storage/azure-storage-blob/azure/storage/blob/container_client.py]: I'm guessing you have already discussed with Laurent regarding his updates to the paging implementation and how this interacts with tracing? Is anything required on the paging models themselves?

File: sdk/storage/azure-storage-blob/azure/storage/blob/container_client.py (hunk 5 above)
Comment (SuyogSoti) [reply sdk/storage/azure-storage-blob/azure/storage/blob/container_client.py]: so the idea is that there should only be two spans. The span for the entry point function and the span for the network. Is the customer ever expected to call aquire_lease? If not I think the best thing to do would be to just not decorate any of them

File: sdk/storage/azure-storage-blob/azure/storage/blob/container_client.py (hunk 7 above)
Comment (SuyogSoti) [reply sdk/storage/azure-storage-blob/azure/storage/blob/container_client.py]: So for paging, the agreement was that the list_* will get the span and each network call will be traced. The function that does the individual network call for paging will not be traced as of preview 2.

File: sdk/storage/azure-storage-blob/azure/storage/blob/container_client.py (hunk 4 above)
Comment (SuyogSoti) [reply sdk/storage/azure-storage-blob/azure/storage/blob/container_client.py]: ok decorator is no longer there

File: sdk/storage/azure-storage-blob/azure/storage/blob/blob_service_client.py (hunk 3 above)
Comment (SuyogSoti) [reply sdk/storage/azure-storage-blob/azure/storage/blob/blob_service_client.py]: ok decorator is no longer there

File: sdk/storage/azure-storage-blob/azure/storage/blob/blob_client.py (hunk 2 above)
Comment (SuyogSoti) [reply sdk/storage/azure-storage-blob/azure/storage/blob/blob_client.py]: ok decorator is no longer there

File: sdk/storage/azure-storage-blob/azure/storage/blob/container_client.py (hunk 5 above)
Comment (annatisch) [reply sdk/storage/azure-storage-blob/azure/storage/blob/container_client.py]: Yes the customer is expected to call this - it's just that there's two paths to access the same function:
a) Customer directly calls Lease.aquire_lease
b) Customer calls Blob or ContainerClient.aquire_lease, which then calls Lease.aquire_lease.

Currently all three of these functions are decorated - but only one of them makes a network call.

File: sdk/storage/azure-storage-blob/azure/storage/blob/container_client.py (hunk 6 above)
Comment (SuyogSoti) [reply sdk/storage/azure-storage-blob/azure/storage/blob/container_client.py]: ok so I talked to Johan and it is his opinion that there will need to be some refactoring in the future: #6492

File: sdk/storage/azure-storage-blob/azure/storage/blob/container_client.py (hunk 5 above)
Comment (SuyogSoti) [reply sdk/storage/azure-storage-blob/azure/storage/blob/container_client.py]: I think this is the same problem as the `upload_blob` problem: #6492.

**Pull request merged**